staging/src/k8s.io/apiserver/plugin/pkg/audit/webhook/webhook.go

```diff
@@ -0,0 +1,215 @@
+	case ModeBlocking:
+	default:
+		return nil, fmt.Errorf("webhook mode %q is not in list of known modes (%s)",
+			mode, strings.Join(AllowedModes, ","))
+	}
+
+	return b, nil
+}
+
+// Run starts the webhook and shuts down once stopCh closes. It MUST be called before
+// using the backend. Run is non-blocking.
+func (a *AuditBackend) Run(stopCh <-chan struct{}) error {
+	if a.buffer == nil {
+		// Not buffered, nothing to do.
+		return nil
+	}
+
+	go func() {
+		for {
+			t := time.NewTicker(a.maxBatchWait)
+			keepGoing, err := a.sendBatchEvents(stopCh, t.C)
+			t.Stop() // Release ticker resources
+
+			if err != nil {
+				glog.Errorf("failed to POST webhook events: %v", err)
+			}
+			if !keepGoing {
+				// stopCh as been closed.
+				return
+			}
+		}
+	}()
+	return nil
+}
+
+// ProcessEvents attempts to POST a serialized audit event to an external service. The actual
+// payload contains an audit.EventList with a single element. The audit webhook may batch
+// multiple events in the future.
+func (a *AuditBackend) ProcessEvents(ev ...*auditinternal.Event) {
+	if a.buffer != nil {
+		a.queueEvents(ev...)
+		return
+	}
+
+	if err := a.sendEvents(ev...); err != nil {
+		glog.Errorf("failed to POST webhook events: %v", err)
+	}
+}
+
+// queueEvents adds the provided events to the internal buffer.
+func (a *AuditBackend) queueEvents(ev ...*auditinternal.Event) {
+	for _, e := range ev {
+		a.buffer <- e
+	}
+}
+
+// sendBatchEvents attempts to batch some number of events to the backend. It returns any errors
+// experienced when trying to POST events and a boolean indicating if the webhook has been shut down.
+//
+// The following things can cause sendBatchEvents to exit:
+//
+//   * Some maximum number of events are received and sent.
+//   * Timer has passed, all queued events are sent.
+//   * StopCh is closed, all queued evnets are sent and keepGoing is false.
+//
+func (a *AuditBackend) sendBatchEvents(stopCh <-chan struct{}, timer <-chan time.Time) (keepGoing bool, err error) {
+	var events []auditinternal.Event
+	keepGoing = true
+
+	for i := 0; i < a.maxBatchSize; i++ {
+		select {
+		case <-timer:
+			// Timer has expired. Send whatever events are in the queue.
+			continue
+		case <-stopCh:
+			// Webhook has shut down. Send the last events and shut down.
+			keepGoing = false
+			continue
+		case ev := <-a.buffer:
+			events = append(events, *ev)
+		}
+	}
+
+	if len(events) == 0 {
+		// No events received.
+		return keepGoing, nil
+	}
+
+	list := auditinternal.EventList{Items: events}
+	return keepGoing, webhook.WithExponentialBackoff(0, func() error {
+		return a.w.RestClient.Post().Body(&list).Do().Error()
+	})
+}
+
+// sendEvents sends a single POST to the backend including all the passed events.
+func (a *AuditBackend) sendEvents(ev ...*auditinternal.Event) error {
+	var list auditinternal.EventList
+	for _, e := range ev {
+		list.Items = append(list.Items, *e)
+	}
+	return a.w.RestClient.Post().Body(&list).Do().Error()
```

Comment:
This is the blocking version. Exponential backoff seems riskier since there's no buffering.